SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -0,0 +1,539 @@
+import SSA.Core.MLIRSyntax.EDSL
+import SSA.Projects.CIRCT.Stream.Stream
+import SSA.Projects.CIRCT.Stream.WeakBisim
+
+
+namespace Comb
+
+
+def add {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.add (0#w) l
+
+def and {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.and (0#w) l
+
+/-- Concatenate a list of bitvecs `xs`, where the length of bitvec xs[i] is given by
+  element ls[i] in a list of nat `ls`-/
+def concat {ls : List Nat} (xs : HVector BitVec ls) : BitVec (List.sum ls) :=
+  match (xs) with
+  | (.nil) => 0#0
+  | (.cons x xs) =>
+     let xsSum := concat xs
+     x ++ xsSum
+
+def divs (x y : BitVec w) : BitVec w :=
+  BitVec.sdiv x y
+
+def divu (x y : BitVec w) : BitVec w :=
+  BitVec.udiv x y
+
+def extract (x : BitVec w) (lb : Nat) : BitVec (w - lb) :=
+  BitVec.truncate (w - lb) (BitVec.ushiftRight x lb)
+
+def icmp (x y : BitVec w) (op : Nat) : Bool :=
+  match op with
+  | 0 => x == y
+  | 1 => x != y
+  | 2 => BitVec.slt x y
+  | 3 => BitVec.sle x y
+  | 4 => x.toInt > y.toInt
+  | 5 => x.toInt ≥  y.toInt
+  | 6 => BitVec.ult x y
+  | 7 => BitVec.ule x y
+  | 8 => x.toNat > y.toNat
+  | 9 => x.toNat > y.toNat
+  /- ceq/cne = SV case in/equality ("===")
+    tests 4-state logical in/equality (0, 1, x, z)
+    TODO: model the concept.
```

Comment:
Rather than say `TODO: model the concept` I'd add a comment here to clarify that we currently only model the 0 and 1 states, not the full 4-valued logic. It's not clear what 'the concept' in this note refers to.

If you really want the todo, I'd add it in the TyDenote class (which presumably denotes into `bitvec w`) as a "TODO: model `x` and `z` states for the full 4-valued logic" or some such